File: Dijkstra/[BOJ]1916_최소비용 구하기/1916.js
const fs = require('fs')
const filePath = process.platform === 'linux' ? 'dev/stdin' : '1916_input.txt'
const input = fs.readFileSync(filePath).toString().trim().split('\n')

const N = Number(input[0])  // 도시의 개수
const M = Number(input[1])  // 버스의 개수

const graph = Array.from({ length: N + 1 }, () => []);

for (let i=2; i < 2+M; i++) {
  const [s, e, c] = input[i].split(' ').map(Number)
  graph[s].push([e, c])
}
const LIMIT = 100001
const distance = Array(N+1).fill(LIMIT)  // 도시 

const [start, end] = input[2+M].split(' ').map(Number)

const dijkstra = x => {

  distance[x] = 0
  
  const q = [[x, 0]]
  let front = 0

  while ( front < q.length ) {
    const [cur, curDist] = q[front++]

    if ( distance[cur] < curDist ) continue

    for (const [nxt, nxtDist] of graph[cur]) {
      const newDist = curDist + nxtDist 
      if ( newDist >= distance[nxt] ) continue

      distance[nxt] = newDist
      q.push([nxt, newDist])
    }
  }

} 

dijkstra(start)
console.log(distance[end])

File: Dijkstra/[BOJ]1916_최소비용 구하기/1916.py
import sys
input = sys.stdin.readline

from heapq import heappop, heappush


def dijkstra (v) :

    cost_table[v] = 0
    heap = []
    heappush(heap, [v, 0])  # [node, cost]
    while heap:
        cur, cur_cost = heappop(heap)

        if cost_table[cur] < cur_cost:
            continue

        for nxt, nxt_cost in graph[cur]:
            sum_cost = cur_cost + nxt_cost

            if cost_table[nxt] <= sum_cost:
                continue

            cost_table[nxt] = sum_cost
            heappush(heap, [nxt, sum_cost])

N = int(input()) # 도시의 개수
M = int(input()) # 버스의 개수

graph = [[] for i in range(N+1)]
for i in range(M):
    start, end, cost = map(int, input().split())
    graph[start].append([end, cost])

start_pnt, end_pnt = map(int, input().split())  # 시작지점, 끝지점

INIT = 100000 * N
cost_table = [INIT] * (N + 1)
dijkstra(start_pnt)

print(cost_table[end_pnt])
